src/openssl/ssl/hostname_validation.cr

```diff
@@ -0,0 +1,154 @@
+    result
+  end
+
+  # Matches hostname against Subject Alternate Name (SAN) entries of certificate.
+  #
+  # Adapted from https://wiki.openssl.org/index.php/Hostname_validation
+  def self.matches_subject_alternative_name(hostname, server_cert : LibCrypto::X509)
+    san_names = LibCrypto.x509_get_ext_d2i(server_cert, LibCrypto::NID_subject_alt_name, nil, nil)
+    return Result::NoSANPresent if san_names.null?
+
+    LibCrypto.sk_num(san_names).times do |i|
+      current_name = LibCrypto.sk_value(san_names, i).as(LibCrypto::GENERAL_NAME*).value
+      next unless current_name.type == LibCrypto::GEN_DNS
+
+      dns_name = LibCrypto.asn1_string_data(current_name.dns_name)
+      dns_name_len = LibCrypto.asn1_string_length(current_name.dns_name)
+      return Result::MalformedCertificate if dns_name_len != LibC.strlen(dns_name)
+
+      pattern = String.new(dns_name, dns_name_len)
+      return Result::MatchFound if matches_hostname?(pattern, hostname)
+    end
+
+    Result::MatchNotFound
+  ensure
+    LibCrypto.sk_pop_free(san_names, ->(ptr : Void*) {
+      LibCrypto.sk_free(ptr)
+    })
+  end
+
+  # Matches hostname from Common Name (CN) entry of certificate. Should only be
+  # called if no SAN entries could be found in certificate.
+  #
+  # Adapted from https://wiki.openssl.org/index.php/Hostname_validation
+  def self.matches_common_name(hostname, server_cert : LibCrypto::X509)
+    index = LibCrypto.x509_name_get_index_by_nid(LibCrypto.x509_get_subject_name(server_cert), LibCrypto::NID_commonName, -1)
+    return Result::Error if index < 0
+
+    name_entry = LibCrypto.x509_name_get_entry(LibCrypto.x509_get_subject_name(server_cert), index)
+    return Result::Error if name_entry.null?
+
+    asn1 = LibCrypto.x509_name_entry_get_data(name_entry)
+    return Result::Error if asn1.null?
+
+    str = LibCrypto.asn1_string_data(asn1)
+    str_len = LibCrypto.asn1_string_length(asn1)
+    return Result::MalformedCertificate if str_len != LibC.strlen(str)
+
+    common_name = String.new(str, str_len)
+    return Result::MatchFound if matches_hostname?(common_name, hostname)
+
+    Result::MatchNotFound
+  end
+
+  # Matches a hostname against a wildcard pattern.
+  #
+  # The hostname must be an exact match or use a wildcard following
+  # [RFC6125, section 6.4.3](http://tools.ietf.org/html/rfc6125#section-6.4.3)
+  # and [RFC6125, section 7.2](http://tools.ietf.org/html/rfc6125#section-7.2)
+  #
+  # In addition, no wildcard match is attempted for IP addresses, and the
+  # hostname and patterns are normalized to skip trailing dots (like browsers
+  # do).
+  #
+  # Adapted from cURL:
+  # Copyright (C) 1998 - 2014, Daniel Stenberg, <daniel@haxx.se>, et al.
+  # https://github.com/curl/curl/blob/curl-7_41_0/lib/hostcheck.c
+  def self.matches_hostname?(pattern, hostname)
+    pattern = pattern.chomp('.').downcase
+    hostname = hostname.chomp('.').downcase
+
+    if ::Socket.ip?(hostname)
```

Comment:
This is wrong: this should be in `matches_subject_alternative_name` since the SAN will be either DNS (2) or IP (7), and I only take DNS (ie. domain name), and CN only contains a domain name.